Modules/_abc.c

```diff
@@ -0,0 +1,894 @@
+/* ABCMeta implementation */
+
+/* TODO: Think (ask) about thread-safety. */
+/* TODO: Add checks only to those calls that can fail and use _GET_SIZE etc. */
+/* TODO: Think about inlining some calls (like __subclasses__) and/or using macros */
+/* TODO: Use separate branches with "fast paths" */
```

Comment:
> TODO: Think (ask) about thread-safety.

What about it?  We have GIL, so C API is protected.  If Python ABCMeta didn't use locks, then you shouldn't too.

> TODO: Think about inlining some calls (like __subclasses__) and/or using macros
> TODO: Use separate branches with "fast paths"

I honestly don't expect these optimizations to yield any detectable speedup so I'd suggest you to try this only if you have time.

Reply:
OK, I just wanted to be sure that our solution for the tricky part about deletion during iteration will not be broken in multi-threaded context.

Btw, we use normal sets for caches (just to make them faster and simpler). But I noticed that we also once call  `.copy()` on them. From brief reading through the `set` code I am not 100% sure it is safe (it doesn't directly iterate over set, but I am not familiar with `set` internals). In case it is not safe, I would rather don't call `.copy()`, this is only used to have nicely printed caches.

Reply:
> OK, I just wanted to be sure that our solution for the tricky part about deletion during iteration will not be broken in multi-threaded context.

Oh, this might be tricky.  I'd have to use a lock for the iteration :(

The way I'd approach this is to try to break it first -- write a fuzzing unittest and try to crash/break the code.  Once you have the test we can discuss potential solutions.  But let's get the rest of the patch in shape first.

Reply:
> [..] I would rather don't call .copy(), this is only used to have nicely printed caches.

We use `.copy` only in the `_get_dump` debug function, right? If so, I think it's fine regardless of how `set.copy()` works.

Reply:
> We use .copy only in the `_get_dump` debug function, right?

Yes.

> The way I'd approach this is to try to break it first.

My hope is that I tried to mimic the current Python version as close as possible, so at least it will be not easy to break it. (For example the case where multiple iterators are iterating the same set seems to be already covered.)